src/utils/filter-utils.js

```diff
@@ -268,110 +272,181 @@ export function updateFilterField(filter, idx, dataset, filters) {
+    const dynamicDomainFilters = shouldCalDomain ? filterRecord.dynamicDomain : null;
+    const cpuFilters = shouldCalIndex ? (opt.cpuOnly ? filters : filterRecord.cpu) : null;
+    filterResult = filterDataByFilterTypes(
+      {dynamicDomainFilters, cpuFilters},
+      allData
     );
+  }
+
+  return {
+    ...newDataset,
+    ...filterResult,
+    gpuFilter: getGpuFilterProps(filters, dataId)
+  };
+}
+
+/**
+ *
+ * @param {Object} filters
+ * @param {Array|null} filters.dynamicDomainFilters
+ * @param {Array|null} filters.cpuFilters
+ * @returns {{filteredIndex: Array, filteredIndexForDomain: Array}} filteredIndex and filteredIndexForDomain
+ */
+function filterDataByFilterTypes({dynamicDomainFilters, cpuFilters}, allData) {
+  // const filteredIndexForDomain = [];
+  // const filteredIndex = [];
+  const result = {
+    ...(dynamicDomainFilters ? {filteredIndexForDomain: []} : {}),
+    ...(cpuFilters ? {filteredIndex: []} : {})
+  };
+
+  for (let i = 0; i < allData.length; i++) {
+    const matchForDomain =
+      dynamicDomainFilters &&
+      dynamicDomainFilters.every(filter =>
+        isDataMatchFilter(allData[i], filter, i)
+      );
 
     if (matchForDomain) {
-      filteredIndexForDomain.push(i);
+      result.filteredIndexForDomain.push(i);
+    }
 
-      // filter data for render
-      const matchForRender = fixedDomainFilters.every(filter =>
-        // if gpu filter, no need to filter data
-        isDataMatchFilter(d, filter, i)
+    const matchForRender =
+      cpuFilters &&
+      cpuFilters.every(filter =>
+        isDataMatchFilter(allData[i], filter, i)
       );
 
-      if (matchForRender) {
-        filteredIndex.push(i);
-      }
+    if (matchForRender) {
+      result.filteredIndex.push(i);
     }
   }
 
-  return {filteredIndex, filteredIndexForDomain};
+  return result;
 }
 
+/**
+ * Get a record of filters based on domain type and gpu / cpu
+ * @param {string} dataId
+ * @param {Array<Object>} filters
+ * @returns {{dynamicDomain: Array, fixedDomain: Array, cpu: Array, gpu: Array}} filterRecord
+ */
+export function getFilterRecord(dataId, filters) {
+  const filterRecord = {
+    dynamicDomain: [],
+    fixedDomain: [],
+    cpu: [],
+    gpu: []
+  };
+
+  filters.forEach(f => {
+    if (f.dataId === dataId && f.fieldIdx > -1 && f.value !== null) {
+      (f.fixedDomain
+        ? filterRecord.fixedDomain
+        : filterRecord.dynamicDomain
+      ).push(f);
+      (f.gpu ? filterRecord.gpu : filterRecord.cpu).push(f);
+    }
+  });
+
+  return filterRecord;
+}
+
+/**
+ * Compare filter records to get what has changed
+ * @param {Object} filterRecord
+ * @param {Object} oldFilterRecord
+ * @returns {{dynamicDomain: Object, fixedDomain: Object, cpu: Object, gpu: Object}} changed filters based on type
+ */
+export function diffFilters(filterRecord, oldFilterRecord = {}) {
+  let filterChanged = {};
+
+  /* eslint-disable no-loop-func */
+  for (const record in filterRecord) {
```

Comment:
if you want you can combine Object.entries and reduce to retrieve keys and values of a given object

Reply:
updated